Add initial accordion block

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This is a copy of https://github.com/WordPress/gutenberg/pull/64119 with a couple changes:
- Moved `render` to the class creation that uses the `AbstractBlock`
- Added the `role="region"` to the panel container ( as suggested in the ARIA docs -> https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/#rps_label )
- Updated interactivity API to use the Woo version for now ( had to rename the `view.js` file to `frontend.js` because of this ).
- Removed the `"style": "wp-block-accordion-group"` and managed that as part of the `AccordionBlock` class.
- Renamed block namespace from `core` to `woocommerce`.
- Updated some of the **WordPress dependencies** doc comments to **External dependencies**.
- Updated the Webpack config to not bundle `parent` blocks within a `inner-blocks` folder just for the Accordion block.

Initial part of #54666 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install the [WooCommerce Beta Tester plugin](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-beta-tester)
2. Add a new page or post and search for **Accordion** in the block inserter ( nothing should show ).
3. Go to tools **Tools > WCA Test Helper > Features** and enable the `experimental-blocks` feature flag
4. Go back to the new page or post and search for **Accordion** in the block inserter.
5. **Accordion group** should show up, add this, this should automatically add two accordions with a header and panel each.
6. Add some content to the header and to the panel, to one of the panels add a link.
7. Save the page/post and go to preview, the accordions should be closed by default ( you can change this with the **Open by default** attribute on the accordion ).
8. Open each accordion and close again, it should work correctly.
9. Use your keyboard, you should be able to tab through the accordion headers and open the accordion when the header is focused using _Space_ or _Enter_ ( see ARIA requirements: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/ )
10. When an accordion is open and contains focusable content ( like a link ) clicking tab should correctly navigate into the panel and select the content ( link ). When the accordion is closed it should not do this.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Add initial copied accordion related blocks

louwie17 committed:
Added Block classes for Accordion block and updated to use Woo interactivity API

louwie17 committed:
Add warning and delete init.js file

gigitux commented on plugins/woocommerce-blocks/assets/js/blocks/accordion-group/edit.js:
I would suggest adding a comment that mentions that we are copy-pasting this block, linking to the original PR.

gigitux replied:
Maybe a README could be a better idea? 

I would change the file folder structure to:

- \accordion
   - \inner-blocks
       -  \accordion-group
       - \accordion-header
       - \accordion-item
        - \accordion-panel
   - README.md

louwie17 replied:
Done as part of: https://github.com/woocommerce/woocommerce/pull/54684/commits/c32ee8c2a904ccf9e6b599852e5f8445dcee4fd2
I also moved the PHP classes into their own folder in this commit: https://github.com/woocommerce/woocommerce/pull/54684/commits/d831b12e0e87c98461330de80288974f87912ea8

gigitux commented on plugins/woocommerce-blocks/assets/js/blocks/accordion-header/index.js:
```diff
@@ -0,0 +1,47 @@
+/**
+ * WARNING: This block is a copy of an upstream implementation: https://github.com/WordPress/gutenberg/pull/64119
+ * Please keep changes to a minimum.
```
Ah cool, I just saw that you added on all the `index.js` files. I don't have a strong preference, but changing the folder structure and adding a README.md could be more aligned with our folder structure (see product gallery or product collection)

louwie17 replied:
I removed these warnings and only left one in the README.
Let me know if you do think we should add something like the above in the `edit` and `save` files?

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionGroup.php:
```diff
@@ -0,0 +1,61 @@
+		}
+	
+		$suffix = wp_scripts_get_suffix();
+		if ( defined( 'IS_GUTENBERG_PLUGIN' ) && IS_GUTENBERG_PLUGIN ) {
+			$module_url = gutenberg_url( '/build-module/block-library/accordion-group/view.min.js' );
+		}
```
Is this necessary?

louwie17 replied:
No, I can remove it!
Although if the accordion does end up being merged to GB, we did have to implement something similar I think.

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionGroup.php:
```diff
@@ -0,0 +1,61 @@
+		}
+	
+		$script_data = $this->asset_api->get_script_data( $this->asset_api->get_block_asset_build_path( $this->block_name . '-frontend' ), array(  ) );
+		wp_register_script(
+			'wc-accordion-group-frontend',
+			$script_data['src'],
+			$script_data['dependencies'],
+			$script_data['version']
+		);
```
Could we use `get_block_type_script` function? 
This is an example:
https://github.com/woocommerce/woocommerce/blob/4fb757d25df3050a71532a20b0dc17dd8a16c1a5/plugins/woocommerce/src/Blocks/BlockTypes/ProductButton.php#L21-L35

louwie17 replied:
Thanks, turns out I could remove it all together.

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionGroup.php:
```diff
@@ -0,0 +1,61 @@
+			$script_data['version']
+		);
+	
+		wp_enqueue_script( 'wc-accordion-group-frontend' );
```
I think that we can avoid this, if we follow the approach above.

louwie17 replied:
Done.

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionHeader.php:
```diff
@@ -0,0 +1,34 @@
+	 * @return string[]|null
+	 */
+	protected function get_block_type_style() {
+		return [ 'wc-blocks-style' ];
```
Is it necessary? I think that the CSS that we added doesn't depend on `wc-blocks-style`. For this reason, I think that we can return `null`.

louwie17 replied:
Done as part of https://github.com/woocommerce/woocommerce/pull/54684/commits/8bd13083d60cac5ba12e07cae4e7d4a5c70691d0

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionItem.php:
```diff
@@ -0,0 +1,92 @@
+	 * @return string[]|null
+	 */
+	protected function get_block_type_style() {
+		return [ 'wc-blocks-style' ];
```
Same here

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionHeader.php:
```diff
@@ -0,0 +1,34 @@
+	 * @return array|string|null
+	 */
+	protected function get_block_type_script( $key = null ) {
+        return null;
```
You missed two spaces

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionItem.php:
```diff
@@ -0,0 +1,92 @@
+	 * @return array|string|null
+	 */
+	protected function get_block_type_script( $key = null ) {
+        return null;
```
Same here

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/AccordionPanel.php:
```diff
@@ -0,0 +1,34 @@
+	 * @return string[]|null
+	 */
+	protected function get_block_type_style() {
+		return [ 'wc-blocks-style' ];
```
Same here

gigitux reviewed (commented):
Thanks for working on this! Awesome work! I left a few comments, but nothing so complex to change/fix!

louwie17 committed:
Use wc_initial_state

louwie17 committed:
Clean up classes for Accordion

louwie17 committed:
Update Accordion class structure

louwie17 committed:
Update folder structure of accordion block JS files

louwie17 committed:
Updating spacing

louwie17 committed:
Fix boolean usage in context

louwie17 committed:
Add README

louwie17 committed:
Remove warning

louwie17 committed:
Address styling lint

louwie17 committed:
Add region and prefix unique id with woocommerce

louwie17 committed:
Add changelog

louwie17 committed:
Fix lint errors

louwie17 commented:
@gigitux thanks for the initial review, I addressed all your suggestions. This should be ready for a re-review.
I also added testing instructions for a formal review :)

louwie17 committed:
Remove addition of new lines

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Accordion/AccordionGroup.php:
```diff
@@ -0,0 +1,49 @@
+			return $content;
+		}
+	
+		$suffix = wp_scripts_get_suffix();
```
I think that we can remove this line

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Accordion/AccordionGroup.php:
```diff
@@ -0,0 +1,49 @@
+	 * @return string Rendered block type output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		if ( ! $content ) {
+			return $content;
+		}
```
Is there a particular reason for this check?

gigitux replied:
I just saw that it is present in the Gutenberg too! Let's keep it!

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Accordion/AccordionHeader.php:
```diff
@@ -0,0 +1,36 @@
+	 * @return array|string|null
+	 */
+	protected function get_block_type_script( $key = null ) {
+    	return null;
```
Missing a space

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Accordion/AccordionItem.php:
```diff
@@ -0,0 +1,94 @@
+	protected function render( $attributes, $content, $block ) {
+			if ( ! $content ) {
+				return $content;
+			}
```
Same here

gigitux replied:
I just saw that it is present in the Gutenberg too! Let's keep it!

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Accordion/AccordionItem.php:
```diff
@@ -0,0 +1,94 @@
+	 * @return array|string|null
+	 */
+	protected function get_block_type_script( $key = null ) {
+    	return null;
```
Missing a space

gigitux reviewed (commented):
Thanks for addressing the feedback! I left few comments, but they are small things.

louwie17 committed:
Address PHP lint

louwie17 committed:
Fix formatting

gigitux reviewed (approved):
Awesome work! LGTM! :shipit:
